# 실기 대비
# 실전 문제풀이
# set 2

## 1) 데이터 및 시나리오

### 갤럭시 사용자 데이터 분석

> 디지털 기기 취급 전문업체 DigiSales는 오프라인 매장에서 갤럭시 제품군을 구매한 이력이 있는 사람들의 단말기 이용 행태를 분석하기 위해 통신사 MT로 부터 익명화 처리된 고객 정보를 일부 구매하였다.  
> 해당 데이터에서 인사이트를 뽑기 위해 다양한 분석을 해보고자 한다.

### 데이터 개요

| 파일명 | 행 | 열 | 인코딩 |
|---|---:|---:|---|
| `galaxy_users.csv` | 7032 | 21 | UTF-8 |

## 1) 데이터 및 시나리오

### 변수 상세

| 변수명 | 유형 | 설명 |
|---|---|---|
| `customerID` | string | 고객 식별자 |
| `gender` | string | 성별 |
| `SeniorCitizen` | int | 시니어 여부 `(1: 시니어, 0: 시니어 아님)` |
| `Partner` | string | 배우자 존재 여부 `(Yes: 있음, No: 없음)` |
| `Dependents` | string | 부양가족 `(Yes: 있음, No: 없음)` |
| `tenure` | int | 현 직장 근속 월수 |
| `PhoneService` | string | 단말 서비스 사용 여부 `(Yes: 사용 중, No: 해지)` |
| `MultipleLines` | string | 다중 회선 개통 여부 `(Yes: 사용 중, No: 미사용)` |
| `InternetService` | string | 인터넷 사용 `(DSL: 저속 서비스, Fiber optic: 고속 서비스, No: 안함)` |
| `OnlineSecurity` | string | 온라인 보안 서비스 `(Yes: 사용 중, No: 미사용)` |
| `OnlineBackup` | string | 온라인 백업 서비스 `(Yes: 사용 중, No: 미사용)` |

### 변수 상세

| 변수명 | 유형 | 설명 |
|---|---|---|
| `DeviceProtection` | string | 기기 보호 서비스 `(Yes: 사용 중, No: 미사용)` |
| `TechSupport` | string | 기술 지원 서비스 `(Yes: 사용 중, No: 미사용)` |
| `StreamingTV` | string | 스트리밍 지원 TV `(Yes: 사용 중, No: 미사용)` |
| `StreamingMovies` | string | 영화 스트리밍 서비스 `(Yes: 사용 중, No: 미사용)` |
| `Contract` | string | 계약 갱신 주기 |
| `PaperlessBilling` | string | 온라인 명세서 발급 서비스 `(Yes: 사용 중, No: 미사용)` |
| `PaymentMethod` | string | 납부 방법 |
| `MonthlyCharges` | float | 월별 청구 금액 |
| `TotalCharges` | float | 누적 청구 금액 |
| `Churn` | string | 이탈 여부 `(Yes: 이탈, No: 잔존)` |

## 2) 문제

### 필요 라이브러리 함수 및 클래스 목록

| 목록 |
|---|
| `from sklearn.preprocessing import MinMaxScaler` |
| `from sklearn.model_selection import train_test_split` |
| `from sklearn.linear_model import LogisticRegression` |
| `from sklearn.metrics import f1_score` |

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score

In [4]:
df = pd.read_csv('../../dataset/galaxy_users.csv')
#SCDI
display(df.shape)
display(df.columns)
display(df.dtypes)
display(df.isna().sum())

(7032, 21)

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

### Q01.

부가 서비스를 많이 사용할수록 제품 또는 브랜드에 고관여 되어있다는 가설을 세우고 분석을 하고자 한다.  
이에 앞서 고객별 부가서비스 사용 현황 통계를 산출하고자 한다.  
부가 서비스를 1개 사용하는 사람은 부가 서비스를 6개 사용하는 사람보다 몇 배나 많은지 계산하시오.

※ 부가서비스 대상 변수: `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`  
※ 부가서비스 변수의 범주가 `"Yes"`인 경우 1로, `"No"`는 0으로 간주한다.  
※ 부가서비스 변수에 `"Yes"` 또는 `"No"`가 아닌 다른 범주가 있는 경우 해당 행은 제거 후 분석하시오.  
※ 첫 번째 고객(`7590-VHVEG`)의 경우 부가서비스는 1개 이용 중이다.  
※ 정답은 반올림하여 소수점 첫째 자리까지 출력하시오. `(정답 예시: 1.2)`

In [46]:
df_q1 = df.copy()
cols_q1 = ['OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']
display(df_q1[cols_q1].apply(lambda x : x.value_counts())) #apply 는 차원을 낮춰서 모두 적용하는 것. dataframe -> 열단위로!
display(df_q1[cols_q1].apply(lambda x : x.str.contains('Yes|No')).sum(axis=1).value_counts()) #No internet service에 no가 있어서 다 포함되는거였다.

#loc로 말고 특정값이 있으면 그 행을 drop한다는 건 없나?
cond_q1 = df_q1[cols_q1].apply(lambda x : x.str.contains('No internet service')).any(axis = 1)
display(cond_q1.value_counts())

df_q1_2 = df_q1.loc[~cond_q1, :].copy()
display(df_q1_2.shape)
df_q1_2[cols_q1] = df_q1_2[cols_q1].replace({
    "Yes" : 1,
    "No"  : 0
}).astype(int) #replace 도 항상 대입해야한다.
display(df_q1_2)

df_q1_2['count'] = df_q1_2[cols_q1].sum(axis = 1)
display(df_q1_2['count'])

service_1 = df_q1_2['count'].value_counts()[1] #열에 대한 value_counts는 ser 니까 index로 접근된다.
service_6 = df_q1_2['count'].value_counts()[6]
round(service_1/service_6, 2)

,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies
No,3497,3087,3094,3472,2809,2781
Yes,2015,2425,2418,2040,2703,2731
No internet service,1520,1520,1520,1520,1520,1520


6    7032
dtype: int64

False    5512
True     1520
dtype: int64

(5512, 21)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,0,...,0,0,0,0,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,1,...,1,0,0,0,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,1,...,0,0,0,0,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,1,...,1,1,0,0,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,0,...,0,0,0,0,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7027,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,1,...,1,1,1,1,One year,Yes,Mailed check,84.80,1990.50,No
7028,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,0,...,1,0,1,1,One year,Yes,Credit card (automatic),103.20,7362.90,No
7029,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,1,...,0,0,0,0,Month-to-month,Yes,Electronic check,29.60,346.45,No
7030,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,0,...,0,0,0,0,Month-to-month,Yes,Mailed check,74.40,306.60,Yes


0       1
1       2
2       2
3       3
4       0
       ..
7027    5
7028    4
7029    1
7030    0
7031    5
Name: count, Length: 5512, dtype: int64

3.4

### Q02.

여러 변수간 상관관계를 확인하고자 한다.  
현 직장 근속 월수(`tenure`), 월별 청구 금액(`MonthlyCharges`), 통신사 사용 월수간 피어슨 상관분석을 실시하고 그 결과 중 상관계수 절대값이 가장 높은 것을 확인하시오.

※ 통신사 사용 월수는 총 청구금액(`TotalCharges`)을 월별 청구금액(`MonthlyCharges`)으로 나눈 값의 몫으로 한다.  
※ 본 문제 풀이에 계약 갱신 주기(`Contract`)는 관련없음.  
※ 정답은 반올림하여 소수점 셋째 자리까지 출력하시오. `(정답 예시: 0.123)`

In [58]:
df_q2 = df.copy()
cols_q2 = ['tenure', 'MonthlyCharges', 'T/M']
df_q2['T/M'] = df_q2['TotalCharges']/df_q2['MonthlyCharges']
display(df_q2[cols_q2].dtypes, df_q2[cols_q2])

ser = df_q2[cols_q2].corr(method='pearson')['tenure'].drop(index='tenure')
#corr() (=dataframe) -> corr()['tenure'] (=series) -> corr()['tenure'].drop은 index를 제거할 수 있다. dataframe은 columns 가 가능하겠지
display(ser.abs().idxmax(), ser.abs().max()) #abs()는 단독으로 쓰면 파라미터 를 절대값으로, series에서는 .abs()로 사용

tenure              int64
MonthlyCharges    float64
T/M               float64
dtype: object

,tenure,MonthlyCharges,T/M
0,1,29.85,1.000000
1,34,56.95,33.178227
2,2,53.85,2.008357
3,45,42.30,43.516548
4,2,70.70,2.144979
...,...,...,...
7027,24,84.80,23.472877
7028,72,103.20,71.345930
7029,11,29.60,11.704392
7030,4,74.40,4.120968


'T/M'

0.9989308049691779

### Q03.

고객 이탈여부(`Churn`)를 종속변수로 하는 모델을 만들어보고자 한다.  
데이터 분할 후 정규화를 실시하여 로지스틱 회귀분석을 실시하고 해당 모델을 기준점 삼아 모델을 개선하고자 한다.  
주어진 독립변수를 활용하여 F1-score로 모델을 평가하시오.

#### 독립변수 / 종속변수

| 구분 | 변수 |
|---|---|
| 독립변수 | 시니어 여부 |
| . | 파트너 존재 여부 |
| . | 부양가족 존재 여부 |
| . | 현 직장 근속 개월 수 |
| . | 월별 청구 금액 |
| . | 누적 청구 금액 |
| . | 온라인 보안 서비스 |
| . | 온라인 백업 서비스 |
| . | 기기 보호 서비스 |
| . | 기술 지원 서비스 |
| . | 영화 스트리밍 서비스 |
| . | 온라인 명세서 발급 서비스 |
| 종속변수 | 고객 이탈여부 |

※ 범주형 변수의 범주가 `"Yes"`인 경우 1로, `"No"`인 경우 0으로 치환하여 분석하시오.  
※ 범주형 변수에서 `"Yes"`와 `"No"`가 아닌 나머지 범주는 -1로 치환하시오.  
※ 학습 및 평가 데이터 세트 분할비는 7:3으로 하시오.  
※ 정규화는 Min-Max 정규화를 실시하며 모든 변수를 대상으로 하시오.  
※ 평가 데이터 세트는 학습 데이터 세트 기반으로 정규화 하시오.  
※ seed는 123으로 고정하시오.  
※ 정답은 반올림하여 소수점 둘째 자리까지 출력하시오. `(정답 예시: 0.12)`

In [ ]:
df_q3 = df.copy()
display(df_q3.columns)
cols_X = ['SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'MonthlyCharges', 'TotalCharges', 'OnlineSecurity', 'OnlineBackup','DeviceProtection', 'TechSupport','StreamingMovies', 'PaperlessBilling']
cols_y = 'Churn'

#D
X = df_q3[cols_X]

def cond(ser) :
    if ser.dtype == 'object' :
        #display(ser)
        display(ser.value_counts())
        ser = ser.replace({
            "Yes" : 1,
            "No"  : 0,
            "No internet service" : -1
        }).astype(int).copy() # 값이 아닌 나머지 범주를 치환하는 가장 쉬운 방법은?
    return ser #이게 중요하다!!!! 리턴이 있어야 apply 밖으로 값이 전달되니까
X = X.apply(cond)
display(X.dtypes, X) #범주형이 없어진것을 확인해야한다. logistic이니까

"""
for col in cols_X :
    if X[col].dtype == 'object' :
        X[col] = X[col].replace({
            "Yes" : 1,
            "No"  : 0,
            "No internet service" : -1
        }).astype(int)
        display(X[col].value_counts())
display(X)
"""

y = df_q3[cols_y]
y = cond(y) #logisticregressor 는 범주형 결과 안됨, 범주형 yes,no 변경은 종속변수도 해당되는 것!!!!
train_X,test_X,train_y,test_y = train_test_split(X,y, train_size=0.7,random_state = 123)

#N
scaler = MinMaxScaler()
train_X_n = scaler.fit_transform(train_X)
test_X_n = scaler.transform(test_X)
#M
model = LogisticRegression(random_state = 123) #왜 random_state가 필요하지
model.fit(train_X_n, train_y)
pred_y = model.predict(test_X_n)
display(test_y, pred_y)
#E
round(f1_score(test_y, pred_y), 2)

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

No     3639
Yes    3393
Name: Partner, dtype: int64

No     4933
Yes    2099
Name: Dependents, dtype: int64

No                     3497
Yes                    2015
No internet service    1520
Name: OnlineSecurity, dtype: int64

No                     3087
Yes                    2425
No internet service    1520
Name: OnlineBackup, dtype: int64

No                     3094
Yes                    2418
No internet service    1520
Name: DeviceProtection, dtype: int64

No                     3472
Yes                    2040
No internet service    1520
Name: TechSupport, dtype: int64

No                     2781
Yes                    2731
No internet service    1520
Name: StreamingMovies, dtype: int64

Yes    4168
No     2864
Name: PaperlessBilling, dtype: int64

SeniorCitizen         int64
Partner               int32
Dependents            int32
tenure                int64
MonthlyCharges      float64
TotalCharges        float64
OnlineSecurity        int32
OnlineBackup          int32
DeviceProtection      int32
TechSupport           int32
StreamingMovies       int32
PaperlessBilling      int32
dtype: object

,SeniorCitizen,Partner,Dependents,tenure,MonthlyCharges,TotalCharges,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingMovies,PaperlessBilling
0,0,1,0,1,29.85,29.85,0,1,0,0,0,1
1,0,0,0,34,56.95,1889.50,1,0,1,0,0,0
2,0,0,0,2,53.85,108.15,1,1,0,0,0,1
3,0,0,0,45,42.30,1840.75,1,0,1,1,0,0
4,0,0,0,2,70.70,151.65,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...
7027,0,1,1,24,84.80,1990.50,1,0,1,1,1,1
7028,0,1,1,72,103.20,7362.90,0,1,1,0,1,1
7029,0,1,1,11,29.60,346.45,1,0,0,0,0,1
7030,1,1,0,4,74.40,306.60,0,0,0,0,0,1


No     5163
Yes    1869
Name: Churn, dtype: int64

6819    0
364     0
2062    0
6953    0
4860    1
       ..
1479    0
3754    0
6604    0
6025    0
3844    0
Name: Churn, Length: 2110, dtype: int32

array([0, 0, 0, ..., 0, 0, 0])

0.55

# 오답노트

## 최종 답

- Q01: **3.4배**
- Q02: **0.999**
- Q03: **0.55**

## Q01 — 유효한 범주만 남기고 서비스 개수 세기

### 1. `contains('Yes|No')`가 틀리는 이유

`str.contains()`는 부분 문자열을 검색한다. 따라서 `No internet service`에도 `No`가 포함되어 `True`가 된다. 이 문제는 셀 전체 값이 정확히 `Yes` 또는 `No`인지 확인해야 하므로 `isin()`을 사용한다.

```python
cols_q1 = [
    'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
    'TechSupport', 'StreamingTV', 'StreamingMovies'
]

valid = df_q1[cols_q1].isin(['Yes', 'No']).all(axis=1)
df_q1_valid = df_q1.loc[valid].copy()
```

- `isin(['Yes', 'No'])`: 각 셀을 정확히 비교
- `.all(axis=1)`: 한 행의 6개 값이 모두 유효한지 확인
- `~valid`: 유효하지 않은 행을 선택할 때 사용

특정 값인 `No internet service`만 제거해도 현재 데이터에서는 맞지만, 문제는 **Yes/No가 아닌 모든 범주**를 제거하라고 했으므로 `isin(...).all(axis=1)`이 더 정확하다. 행을 직접 삭제하고 싶다면 `df.drop(index=df.index[~valid])`도 가능하지만 Boolean 필터링이 간결하다.

### 2. `apply()`의 정확한 의미

`DataFrame.apply()`가 항상 차원을 낮추는 것은 아니다. 기본값 `axis=0`에서는 각 열을 Series로 함수에 전달하고, `axis=1`에서는 각 행을 전달한다. 이 문제의 유효성 검사는 `apply()` 없이 `isin()`으로 한 번에 처리할 수 있다.

### 3. 변환, 행 합계, 빈도 계산

```python
df_q1_valid[cols_q1] = (
    df_q1_valid[cols_q1]
    .replace({'Yes': 1, 'No': 0})
    .astype(int)
)
df_q1_valid['service_count'] = df_q1_valid[cols_q1].sum(axis=1)
service_freq = df_q1_valid['service_count'].value_counts()
answer_q1 = round(service_freq.loc[1] / service_freq.loc[6], 1)
```

`replace()` 결과는 다시 대입해야 한다. `value_counts()` 결과는 Series이므로 서비스 개수 라벨은 `.loc[1]`, `.loc[6]`으로 접근하면 의미가 명확하다. 문제는 소수점 **첫째 자리**까지 요구하므로 `round(..., 1)`을 사용한다.

## Q02 — 몫으로 사용 월수를 만들고 전체 상관관계 비교

### 1. `/`가 아니라 `//` 사용

문제에서 `TotalCharges / MonthlyCharges`의 **몫**을 사용하라고 했으므로 일반 나눗셈 `/`가 아니라 몫 연산자 `//`를 사용한다. `/`를 사용한 결과도 반올림하면 우연히 `0.999`가 나오지만 문제 조건을 만족한 풀이는 아니다.

```python
df_q2 = df[['tenure', 'MonthlyCharges']].copy()
df_q2['used_month'] = df['TotalCharges'] // df['MonthlyCharges']
```

### 2. 세 변수의 모든 쌍을 비교하기

`corr()['tenure']`만 선택하면 tenure와 관련된 두 상관계수만 확인한다. 현재 데이터에서는 최댓값이 tenure와 used_month 사이여서 답은 같지만, 문제는 세 변수 간 결과 중 최댓값을 묻기 때문에 전체 상관행렬의 비대각 원소를 비교하는 풀이가 안전하다.

```python
corr_abs = df_q2.corr(method='pearson').abs()
np.fill_diagonal(corr_abs.values, 0)
answer_q2 = round(corr_abs.max().max(), 3)
```

자기 자신과의 상관계수는 항상 1이므로 반드시 제외한다. `np.fill_diagonal()`에는 DataFrame 자체가 아니라 값 배열인 `.values`를 전달한다. 파이썬 숫자는 `abs(x)`, Series/DataFrame은 `.abs()`를 사용할 수 있다.

### 3. 실제 데이터로 `corr_abs.max().max()` 이해하기

`used_month = TotalCharges // MonthlyCharges`로 계산하면 일부 행은 다음과 같다.

| index | tenure | MonthlyCharges | used_month |
|---:|---:|---:|---:|
| 0 | 1 | 29.85 | 1 |
| 1 | 34 | 56.95 | 33 |
| 2 | 2 | 53.85 | 2 |
| 3 | 45 | 42.30 | 43 |
| 7028 | 72 | 103.20 | 71 |
| 7031 | 66 | 105.65 | 64 |

`tenure`와 `used_month` 값이 비슷하게 움직이므로 두 변수의 상관관계가 매우 높을 것으로 예상할 수 있다. 실제 피어슨 상관행렬은 다음과 같다.

|  | tenure | MonthlyCharges | used_month |
|---|---:|---:|---:|
| tenure | 1.000000 | 0.246862 | 0.998831 |
| MonthlyCharges | 0.246862 | 1.000000 | 0.246164 |
| used_month | 0.998831 | 0.246164 | 1.000000 |

행렬은 대칭이므로 `tenure → used_month`와 `used_month → tenure`의 `0.998831`은 같은 관계가 두 번 표시된 것이다. 대각선의 1은 각 변수를 자기 자신과 비교한 결과다.

```python
corr_abs = df_q2.corr(method='pearson').abs()
```

`.abs()`는 음의 상관계수도 양수로 바꾸어 방향이 아닌 상관관계의 강도를 비교하게 한다. 현재 세 상관계수는 모두 양수이므로 실제 숫자는 바뀌지 않는다.

```python
np.fill_diagonal(corr_abs.values, 0)
```

대각선을 0으로 바꾼 실제 행렬:

|  | tenure | MonthlyCharges | used_month |
|---|---:|---:|---:|
| tenure | 0.000000 | 0.246862 | 0.998831 |
| MonthlyCharges | 0.246862 | 0.000000 | 0.246164 |
| used_month | 0.998831 | 0.246164 | 0.000000 |

첫 번째 `.max()`는 각 열의 최댓값을 구한다.

```python
corr_abs.max()

# tenure            0.998831
# MonthlyCharges    0.246862
# used_month        0.998831
```

두 번째 `.max()`는 위 세 값 중 다시 최댓값을 구한다.

```python
corr_abs.max().max()
# 0.998830875...

round(corr_abs.max().max(), 3)
# 0.999
```

즉 `corr()`는 모든 쌍을 계산하지만 `corr()['tenure']`를 선택하면 tenure가 들어간 관계만 남는다. 모든 쌍의 최댓값은 전체 행렬에서 자기상관을 제거한 뒤 `.max().max()`로 구한다. 최댓값에 해당하는 관계는 `tenure ↔ used_month`이다.

## Q03 — 범주 변환, 분할, 정규화, F1-score

### 1. 독립변수 목록을 정확히 옮기기

첫 번째 풀이에는 문제에서 요구한 `StreamingMovies` 대신 `StreamingTV`가 들어갔다. 변수명이 비슷하므로 문제 표를 보고 목록을 그대로 옮긴 뒤 개수와 이름을 다시 확인한다.

```python
cols_X = [
    'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
    'MonthlyCharges', 'TotalCharges', 'OnlineSecurity',
    'OnlineBackup', 'DeviceProtection', 'TechSupport',
    'StreamingMovies', 'PaperlessBilling'
]
```

### 2. Yes/No 외 모든 범주를 일반적으로 -1로 변환하기

```python
def encode_binary(ser):
    if ser.dtype == 'object':
        return ser.map(lambda x: 1 if x == 'Yes' else 0 if x == 'No' else -1)
    return ser

X = df[cols_X].copy().apply(encode_binary)
y = df['Churn'].map({'Yes': 1, 'No': 0})
```

`apply()`에 전달한 함수는 object 열뿐 아니라 숫자 열에서도 반드시 값을 `return`해야 한다. 변환된 결과를 새 변수에 저장한 뒤 실제 모델에도 그 변수를 사용해야 한다. `X = df[cols_X].copy()`처럼 복사하면 값을 바꿀 때 발생하는 `SettingWithCopyWarning`도 방지할 수 있다.

### 3. `y`는 1차원 Series

`cols_y = ['Churn']`과 `df[cols_y]`를 사용하면 y가 2차원 DataFrame이 되어 경고가 발생한다. 단일 종속변수는 `y = df['Churn']`처럼 Series로 만든다.

### 4. 데이터 분할 후 학습 데이터로만 정규화

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=123
)

scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
```

평가 데이터에 `fit_transform()`을 사용하면 평가 데이터의 정보를 미리 학습하는 데이터 누수가 발생한다. 학습 데이터에만 `fit_transform()`, 평가 데이터에는 같은 scaler의 `transform()`만 사용한다. 문제의 '모든 변수'는 모든 **독립변수**를 의미하며 종속변수 y는 정규화하지 않는다.

### 5. 모델 학습과 평가

```python
model = LogisticRegression()
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)
answer_q3 = round(f1_score(y_test, y_pred), 2)
```

seed 123은 데이터 분할의 `random_state`에 적용한다. 기본 solver에서는 `LogisticRegression(random_state=123)`을 추가해도 결과에 실질적인 차이가 없으므로 기준 답안처럼 생략해도 된다.

## 핵심 암기

1. 정확한 범주 검사는 `isin()`, 부분 문자열 검색은 `str.contains()`를 사용한다.
2. 모든 열이 조건을 만족해야 하면 `.all(axis=1)`, 하나라도 만족하면 `.any(axis=1)`을 사용한다.
3. 몫은 `//`, 일반 나눗셈은 `/`이다.
4. 상관계수 최댓값을 구할 때 자기상관 1을 제외한다.
5. 모델 변수 목록에서 `StreamingMovies`와 `StreamingTV`처럼 비슷한 이름을 주의한다.
6. 변환 함수는 모든 분기에서 `return`해야 하며 변환 결과를 실제 모델 입력에 사용한다.
7. scaler는 학습 데이터로 `fit`, 평가 데이터에는 `transform`만 한다.
8. 단일 종속변수 y는 1차원 Series로 전달한다.